File: NMFD/NMFD.m
function [W, H] = NMFD(V,R,T,Niter,initVal)
% [W, H] = NMFD(V,R,T,Niter,initVal)
%    NMFD as proposed by Smaragdis (Non-negative Matrix Factor
%    Deconvolution; Extraction of Multiple Sound Sources from Monophonic
%    Inputs). KL divergence minimization. The proposed algorithm was
%    corrected.
%Input :   
%   - V : magnitude spectrogram to factorize (is a MxN matrix)
%   - R : number of templates
%   - T : template size (in number of frames in the spectrogram)
%   - Niter : number of iterations
%   - initVal (optional) : initialization for W and H. A structure with
%   fields W (initial value of W) and H (initial value of H).
%Output :
%   - W : time/frequency template (TxMxR array, each template is TxM)
%   - H : activities for each template (RxN matrix)
%
% Copyright (C) 2010 Romain Hennequin



% V : spectrogram MxN
% H : activation RxN
% Wt : spectral template MxR t = 0:T-1
% W : TxMxR


% data size
M = size(V,1);
N = size(V,2);

% initialization
if nargin<5
    H = rand(R,N);
    W = rand(M,R,T);
    One = ones(M,N);
    Lambda = zeros(M,N);
else
    H = initVal.H;
    W = initVal.W;
    One = ones(M,N);
    Lambda = zeros(M,N);
end

% Waitbar
message = ['computing NMFD. iteration : 0/' int2str(Niter)];
h = waitbar(0,message);

for iter =1:Niter
    
    % computation of Lambda
    Lambda(:) = 0;
    for f = 1:M
        for z = 1:R
            v = reshape(W(f,z,:),T,1);
            cv = conv(v,H(z,:));
            Lambda(f,:) = Lambda(f,:) + cv(1:size(Lambda,2));
        end
    end

    
    Halt = H;
    
    Htu = zeros(T,R,N);
    Htd = zeros(T,R,N);
   
    % update of H for each value of t (which will be averaged)
    VonLambda = V./(Lambda + eps);
    
    Hu = zeros(R,N);    
    Hd = zeros(R,N);    
    for z = 1:R
        for f = 1:M
            v = reshape(W(f,z,:),T,1);
            cv = conv(VonLambda(f,:),flipud(v));
            Hu(z,:) = Hu(z,:) + cv(T:T+N-1);
            
            v = reshape(W(f,z,:),T,1);
            cv = conv(One(f,:),flipud(v));
            Hd(z,:) = Hd(z,:) + cv(T:T+N-1);
        end
    end
    
    
    % average along t
    H = H.*Hu./Hd;

    figure(20)
    subplot(211)
    imagesc(db(V));
    axis xy
    ax = caxis;
    title('Original spectrogram')
    xlabel('time')
    ylabel('frequency')
    subplot(212)
    imagesc(db(Lambda));
    axis xy
    caxis(ax);
    title('Reconstructed spectrogram')
    xlabel('time')
    ylabel('frequency')
    
% computation of Lambda

    Lambda(:) = 0;
    for f = 1:M
        for z = 1:R
            v = reshape(W(f,z,:),T,1);
            cv = conv(v,H(z,:));
            Lambda(f,:) = Lambda(f,:) + cv(1:size(Lambda,2));
        end
    end


    mu = 1;
    lambda = 1.02.^(0:T-1)-1;
    lambda(1:3) = 0;
    
    SumTot = sum(W,2);
    
    VonLambda = V./(Lambda + eps);
    
    
    % update of Wt
    for t=0:T-1
       W(:,:,t+1) = W(:,:,t+1).*(  (VonLambda*shiftLR(H,-t)') ./ (One*shiftLR(H,-t)'+ eps + mu*lambda(t+1))  );
    end
    
    message = ['computing NMFD. iteration : ' int2str(iter) '/' int2str(Niter)];
    waitbar(iter/Niter,h,message);
end
close(h)

File: python/NMFD.py
import numpy as np
import sys
from scipy.ndimage.filters import convolve1d

def NMFD(V, R=3 ,T=10, n_iter=50, init_W=[], init_H=[]):
    """
     NMFD(V, R=3 ,T=10, n_iter=50, init_W=None, init_H=None)
        NMFD as proposed by Smaragdis (Non-negative Matrix Factor
        Deconvolution; Extraction of Multiple Sound Sources from Monophonic
        Inputs). KL divergence minimization. The proposed algorithm was
        corrected.
    Input :
       - V : magnitude spectrogram to factorize (is a FxN numpy array)
       - R : number of templates (unused if init_W or init_H is set)
       - T : template size (in number of frames in the spectrogram) (unused if init_W is set)
       - n_iter : number of iterations
       - init_W : initial value for W.
       - init_H : initial value for H.
    Output :
       - W : time/frequency template (FxRxT array, each template is TxF)
       - H : activities for each template (RxN array)

     Copyright (C) 2015 Romain Hennequin
    """

    """
     V : spectrogram FxN
     H : activation RxN
     Wt : spectral template FxR t = 0 to T-1
     W : FxRxT
    """
    eps = np.spacing(1)

    # data size
    F = V.shape[0];
    N = V.shape[1];

    # initialization
    if len(init_H):
        H = init_H
        R = H.shape[0]
    else:
        H = np.random.rand(R,N);

    if len(init_W):
        W = init_W
        R = W.shape[1]
        T = W.shape[2]
    else:
        W = np.random.rand(F,R,T);

    One = np.ones((F,N));
    Lambda = np.zeros((F,N));

    cost = np.zeros(n_iter)
    for it in range(n_iter):
        sys.stdout.write('Computing NMFD. iteration : %d/%d' % (it+1, n_iter));
        sys.stdout.write('\r')
        sys.stdout.flush()

        # computation of Lambda
        Lambda[:] = 0;
        for f in range(F):
            for z in range(R):
                cv = np.convolve(W[f,z,:],H[z,:]);
                Lambda[f,:] = Lambda[f,:] + cv[0:Lambda.shape[1]];

        Halt = H.copy();

        Htu = np.zeros((T,R,N));
        Htd = np.zeros((T,R,N));

        # update of H for each value of t (which will be averaged)
        VonLambda = V/(Lambda + eps);

        cost[it] = (V*log(V/Lambda)-V+Lambda).sum()

        Hu = np.zeros((R,N));
        Hd = np.zeros((R,N));
        for z in range(R):
            for f in range(F):
                cv = np.convolve(VonLambda[f,:],np.flipud(W[f,z,:]));
                Hu[z,:] = Hu[z,:] + cv[T-1:T+N-1];
                cv = np.convolve(One[f,:],np.flipud(W[f,z,:]));
                Hd[z,:] = Hd[z,:] + cv[T-1:T+N-1];

        # average along t
        H = H*Hu/Hd;

        # computation of Lambda

        Lambda[:] = 0;
        for f in range(F):
            for z in range(R):
                cv = np.convolve(W[f,z,:],H[z,:])
                Lambda[f,:] += cv[0:Lambda.shape[1]]

        mu = 0
        constraint = 1.02**np.arange(0,T)-1
        constraint[0:2] = 0

        SumTot = W.sum(axis=1)

        VonLambda = V/(Lambda + eps)


        # update of Wt
        for t in range(T):
            shift_H = np.roll(H,-t).T
            if t>0:
                shift_H[-t:,:]=0

            W[:,:,t] = W[:,:,t] * (  np.dot(VonLambda,shift_H) / (np.dot(One,shift_H) + eps + mu * constraint[t])  );

    return (W, H, cost)